node/src/components/gossiper/gossip_table.rs

```diff
@@ -443,40 +413,17 @@ impl<T: Copy + Eq + Hash + Display> GossipTable<T> {
             return action;
         }
 
-        if let Some(state) = self.paused.get_mut(data_id) {
-            if !state.held_by_us {
-                let _ = state.holders.remove(&peer);
-                debug!(item=%data_id, %peer, "removed peer as a holder of the item");
-            }
-        }
-
         GossipAction::Noop
     }
 
-    /// We have deemed the data not suitable for gossiping further.  If left in paused state, the
-    /// entry will eventually be purged, as for finished entries.
-    pub(crate) fn pause(&mut self, data_id: &T) {
-        if let Some(mut state) = self.current.remove(data_id) {
-            state.in_flight_count = 0;
-            let timeout = Instant::now() + self.finished_entry_duration;
-            let _ = self.paused.insert(*data_id, state);
-            let _ = self.paused_timeouts.push(timeout, *data_id);
+    /// We have deemed the data not suitable for gossiping further.  The entry will be marked as
+    /// `finished` and eventually be purged.
+    pub(crate) fn finish(&mut self, data_id: &T) {
+        if self.current.remove(data_id).is_some() {
+            self.insert_to_finished(data_id);
```

Comment:
And here we are not creating a `GossipAction::AnnounceFinished`?

Reply:
I should probably have commented this, sorry!

We're only calling this method when the gossiper has failed to retrieve the deploy from storage.  That should never happen, but if it does, there's no point in emitting an `AnnounceFinished` since the `BlockProposer` will then also just try to get that deploy from storage.

Replying now though, it seems safer to return a `bool` here (`true` if the item was in `self.currrent`) and just let the gossiper announce it and have the `BlockProposer` fail too.

Reply:
Fixed in [b1e63fe](https://github.com/casper-network/casper-node/pull/1484/commits/b1e63fe46556005af8f4a504cdfe5fb8dd77efeb).

Reply:
I'd actually prefer to treat this case differently and have `AnnounceFinished` mean a _successful_ end of gossiping. We could emit another announcement here, if we want to handle it somewhere else. It seems a bit awkward to me that we'll signal to the block proposer something to the effect of "hey, you can now propose this" when we know that we don't even have this piece of data :grimacing:

Reply:
As discussed offline, I've just added a comment to help clarify things should we want to take a different approach here in the future.  Added in [c8c8c3b](https://github.com/casper-network/casper-node/pull/1484/commits/c8c8c3b61502f81a15bdad55ece0de98ae43d184).